Add implementation guide (#96)
ocapn/ocapn: implementation-guide/Implementation Guide.md

She sets up a promise pair and transmits `op:listen` for the promise she got earlier for the beep, the message like this:

```
<op:listen  <desc:export 2>
            <desc:import-object 3>
            true>
```

After another while she gets another message with the resolution to that promise:
```
<op:deliver-only    <desc:import-object 3>
                    ['fulfill "beeeep!"]>
```

Oh cool! that all worked. Now she has promises though she needs to modify how her `op:abort` works, before it just terminated the session if she got one, but now she has promises she needs to ensure they break if the session ends. She sets up the promise creation code to add her promise to a set of unresolved promises that her `op:abort` handling code can access. When the resolver gets the resolution, it removes itself from the set of unresolved promises. She next makes it so that when an `op:abort` comes in, she iterates through the set sending the resolver objects a break with a network partition error. This should ensure promises are handled correctly.

> Promises might not only be fulfilled through, lets imagine that Alisha sents another message to move forward. Lets assume the robot can't do that, it doesn't really matter why, maybe the robot physically isn't able to move forward for there being an obstical, maybe it's got low power, or some other reason entirely, but the promise she gets back instead of being resolved as fulfilled instead gets broken:

Reviewer: ```suggestion
Promises might not only be fulfilled through, lets imagine that Alisha sents another message to move forward. Lets assume the robot can't do that, it doesn't really matter why, maybe the robot physically isn't able to move forward for there being an obstacle, maybe it's got low power, or some other reason entirely, but the promise she gets back instead of being resolved as fulfilled instead gets broken:
```

<op:deliver-only  <desc:export 4> ['break "Unkown error occured"]>
```

### Stage 3: import/export gc

Alisha looks at her export table and sees the following:

- `0 -> <bootstrap object>`
- `1 -> <bootstrap fetch resolve-me-desc>`
- `2 -> <robot beep resolve-me-desc>`
- `3 -> <beep's vow listen resolve-me-desc>`
- `4 -> <move forward resolve-me-desc>`

Alisha already has the reference to the robot, a promise for the response to the beep, and the robot has sent a resolution to the promise it sent Alisha with its reply, "Beeeep!" But these are a lot of objects still hanging around and being exported by both sides. Alisha decides it's time to implement garbage collection so that she may remove these from her table and it doesn't grow too large, ensuring also that Ben's node doesn't have to hang on to all of these objects as well.

There are two garbage collection operations in OCapN's CapTP, but one of thse operations has to do with questions and answers and Alisha isn't doing anything with those yet. Alisha decides to look at the `op:gc-export` operation which looks like this:

```
<op:gc-export export-pos   ; positive integer
              wire-delta>  ; positive integer
```

> The `export-pos` is the same position that's used for the import position and the wire-delta outstanding references to the object that's been received. OCapN's GC implementation is a collaborative process where by both sides keep track of the number of times a given object has been referenced in CapTP messages. When the importing side no longer needs the object, it should send an `op:gc-export` message, with the `wire-delta` being the outstanding reference count. Each time a reference to an object is received the CapTP session should keep track of this and can emit the `op:gc-export` with this amount and remove it from the count. The for keeping track of the wire-delta is the importing node could no longer need the reference and send a GC operation while at the same time the exporting side might send another message which refers to that object, if the exporting side were then to get this GC operation, the exporting side needs to know if it's safe to actually GC it or not.

Reviewer: ```suggestion
The `export-pos` is the same position that's used for the import position and the wire-delta outstanding references to the object that's been received. OCapN's GC implementation is a collaborative process where by both sides keep track of the number of times a given object has been referenced in CapTP messages. When the importing side no longer needs the object, it should send an `op:gc-export` message, with the `wire-delta` being the outstanding reference count. Each time a reference to an object is received the CapTP session should keep track of this and can emit the `op:gc-export` with this amount and remove it from the count. The reason for keeping track of the wire-delta is the importing node could no longer need the reference and send a GC operation while at the same time the exporting side might send another message which refers to that object, if the exporting side were then to get this GC operation, the exporting side needs to know if it's safe to actually GC it or not.
```

;; Now we're sending a message to that promise
<op:deliver <desc:answer 0>
            []
            1
            <desc:import-object 2>>

;; Finally lets send a message to the promise above
<op:deliver <desc:answer 1>
            []
            2
            <desc:import-object 3>>
```

Alisha wants to support this mechanism so when a CapTP session is initialized she creates two further tables to facilitate this:

- Questions: This is to keep track of a mapping between a remote promise descriptor and the answer position
- Answers: This is used to keep track of position to local promises for received `op:deliver` calls.

Starting with questions Alisha modifies her function which creates the `op:deliver` message to add a unique integer for the `answer-pos` instead of the `false` she used earlier. She adds that integer to the questions table and returns her internal representation for this vow that she calls the remote promise descriptor. She also adds to her function the ability to pass in as the target of the message the remote promise descriptor and in that situation it'll check the answer table and send it to that position wrapped in the `desc:answer` record.

> Alisha modifies both her `op:deliver` and `op:deliver-only` message handling code to check if the `to-desc` field is referring to an object or an answer. In the case it's an answer she looks up in her answers table and finds the local promise object which she sends the messages to so they get deliverd when the promise is fulfilled. Alisha also adds to her `op:deliver` message handling code to check for if `answer-pos` is none-null and if so stores the position in her answers table along with the promise created for the `op:deliver`.

Reviewer: ```suggestion
Alisha modifies both her `op:deliver` and `op:deliver-only` message handling code to check if the `to-desc` field is referring to an object or an answer. In the case it's an answer she looks up in her answers table and finds the local promise object which she sends the messages to so they get delivered when the promise is fulfilled. Alisha also adds to her `op:deliver` message handling code to check for if `answer-pos` is not null or `'false` and if so stores the position in her answers table along with the promise created for the `op:deliver`.
```

![](./handoffs-3party1.webm)

In a high level Alisha will deposit a "gift" which is just a reference to the gallery object on Node C's gift table the session Node A and Node C have together. Alisha will then create a sort of certificate called a `desc:handoff-give` which includes important information from both node A and node C's session and node A and node B's session, as well as the ID of the gift she left at C. Node B can create a connection to node C if it doesn't already have one and create a certificate called `desc:handoff-receive` which includes the `desc:handoff-give` certificate A give B, this certificate it delivers to C and will get in return the gift if all went successfully.

That's a lot to take in so we can look at each step one by one and see what's actually happening. The first thing to do before that is to look at the message Alice sends to Ben, the `desc:handoff-give` is used in place of gallery object on Node C the message conceptually would look like this:

```
<op:deliver-only <reference to Ben (desc:import-object)>  [<desc:handoff-give representing the reference to the gallery>]>
```

#### Depositing the gift

![](./handoffs-3party2.webm)

> This is done by the gifter, in our example Alisha's node, Node A. Each node needs a per-session gifting table which stores a gift ID to some reference to an object or promise. The gift ID is just a positive integer (usually incrementing integer) which is unique to the session which represents the reference being shared. When a handoff is initiated it's to a specific object on a node, not just the Node, the gifting mechanism is providing a secure way to share a specific reference to the receiver. The way a node deposits a gift is on the bootstrap object which has a specific `deposit-gift` method.

Reviewer: Which bootstrap object? The one Alisha got from Node C, the Exporter?

btw the `deposit-gift` method on the bootstrap object is nearly the same as the `provideFor()` on the NonceLocator in CapTP of E.

> Alisha starts implementing this by creating a gift table for each session and an auto incrementing counter to track the gift IDs she's sending. Alisha also implements the `deposit-gift` method on her bootstrap object which takes two arguments a gift ID and a reference and she puts these things into her gift table she's just made. Alisha then starts by adding code to her `op:deliver` and `op:deliver-only` to look through all the arguments being sent and when it encounters a reference to an object initialize a handoff and replace that reference with a `desc:handoff-give`. When a handoff is initialized in Alisha's system, it gets the current next gift ID and sends a message to the deposit gift method on the bootstrap object, this looks like this:

Reviewer: ```suggestion
Alisha starts implementing this by creating a gift table for each session and an auto incrementing counter to track the gift IDs she's sending. Alisha also implements the `deposit-gift` method on her bootstrap object which takes two arguments a gift ID and a reference and she puts these things into her gift table she's just made. Alisha then starts by adding code to her `op:deliver` and `op:deliver-only` to look through all the arguments being sent and when it encounters a reference to an object hosted neither by here node or the node at the captp connection it initializes a handoff and replace that reference with a `desc:handoff-give`. When a handoff is initialized in Alisha's system, it gets the current next gift ID and sends a message to the deposit gift method on the ?Exporter? bootstrap object, this looks like this:
```

```
;; Deposit the object Node C is exporting at position `1` at gift ID `5`
<op:deliver-only <desc:export 0> ['deposit-gift 5 <desc:export 1>]
```
 
#### Creating the `desc:handoff-give` certificate

![](./handoffs-3party3.webm)

Alisha next needs to create the `desc:handoff-give` certificate which she can give to Ben on Node B which will enable him to retrieve the reference Alisha left as a gift on Node C. The certificate has the following structure:

```
<desc:handoff-give receiver-key       ; the public key of the receiver in the gifter <-> receiver session
                   exporter-location  ; OCapN Locator for the exporter
                   session            ; The session ID for the gifter <-> exporter session
                   gifter-side        ; the public key of the gifter in the gifter <-> exporter session
                   gift-id>           ; Positive Integer or zero.
```

> This includes a lot of information, some of which comes from the Gifter's session with the Receiver and some of it comes from the Gifter's session with the Exporter. It's important to keep in mind when thinking about handoffs is that it does not rely on keeping secrets, all messages sent can be public while remaining secure and handoffs are capabilities which only allow the receiver to obtain a reference to the object that the gifter deposits. The burdon for checking the information within the certificates rests with the exporter since it is the Node which has the reference to provide.

Reviewer: ```suggestion
This includes a lot of information, some of which comes from the Gifter's session with the Receiver and some of it comes from the Gifter's session with the Exporter. It's important to keep in mind when thinking about handoffs is that it does not rely on keeping secrets, all messages sent can be public while remaining secure and handoffs are capabilities which only allow the receiver to obtain a reference to the object that the gifter deposits. The burden for checking the information within the certificates rests with the exporter since it is the Node which has the reference to provide.
```

### creating the `desc:handoff-receive` certificate

![](./handoffs-3party4.webm)

The receiver doesn't just get the `desc:handoff-give` and pass it along, it generates its own certificate which includes the signed `desc:handoff-give`. The receiver does this to include some additional information only the receiver has that the exporter will need to securely perform the handoff. The certificate the receiver creates is called a `desc:handoff-receive` and looks like this:

```
<desc:handoff-receive receiving-session  ; The ID of the receiver <-> exporter session
                      receiving-side     ; The public key of the receiver in the receiver <-> exporter session
                      handoff-count      ; positive integer
                      signed-give>       ; desc:sig-envolope containing a desc:handoff-give received from the gifter
```

> We can see the receiver adds some information, but most of it is provided already by the gifter and is included by including that signed certificate along with the `desc:handoff-receive` in the `signed-give` field. One piece of information that's included here is the `handoff-count`, each session has an inrecementing integer starting at zero which reflects how many handoff-receive's they've sent in the session. They also include an integer which reflects how many handoff receive's they've received (with them being the exporter) within the session. This is used to prevent replay attacks, the exporter must only provide the reference once to prevent replay attacks. The last thing to note here is this is also wrapped in a `desc:sig-envolope` with a signature, but this signature is created with the receiver's private key in the gifter <-> receiver session. Again, this might seem a bit strange for two reasons:

Reviewer: ```suggestion
We can see the receiver adds some information, but most of it is provided already by the gifter and is included by including that signed certificate along with the `desc:handoff-receive` in the `signed-give` field. One piece of information that's included here is the `handoff-count`, each session has an inrecementing integer starting at zero which reflects how many handoff-receive's they've sent in the session. They also include an integer which reflects how many handoff receive's they've received (with them being the exporter) within the session. This is used to prevent replay attacks, the exporter must only provide the reference once to prevent replay attacks. The last thing to note here is this is also wrapped in a `desc:sig-envelope` with a signature, but this signature is created with the receiver's private key in the gifter <-> receiver session. Again, this might seem a bit strange for two reasons:
```

1. Again, we're sending the `desc:handoff-receive` to the exporter but using a key from another session, the gifter <-> receiver session.

> 2. Since these keys are per-session and generated anew on each new connection, the exporter normally would have now way to verify this signature. The gifter, however, included the receiver's public key they used within the `handoff-give` (the `receiver-key` field) that they signed. The exporter is able to extract this key from the `desc:handoff-give` and use it to verify the signature on the `handoff-receive` (more on this later in the document).

Reviewer: ```suggestion
2. Since these keys are per-session and generated anew on each new connection, the exporter normally would have no way to verify this signature. The gifter, however, included the receiver's public key they used within the `handoff-give` (the `receiver-key` field) that they signed. The exporter is able to extract this key from the `desc:handoff-give` and use it to verify the signature on the `handoff-receive` (more on this later in the document).
```

```
<op:deliver (desc:export 0)
            ['withdraw-gift
             (desc:sig-envolope
              (desc:handoff-receive
                <ID of the newely created session Ben has with Carol>
                0
                (desc:sig-envolope
                  (desc:handoff-give
                    <Ben's public key for his session with Alisha>
                    (ocapn-node "tcenolezzq7vleywviuvwl74dh2nhs3nf7lun5zuhtjpwhjed5ojw6qd" 'onion #f)
                    <ID of the session Alisha and Carol's node have>
                    <Alisha's public key in her session with Carol>
                    5)
                  <signature Alisha made with her private key in the session with Carol's Node>)))
            ]
           #f
           <desc:import-object 1>>
```

> /If en had had a connection to Carol's Node, Node C, his Node would just have used that instead of creating a new one/

Reviewer: ```suggestion
/If Ben had had a connection to Carol's Node, Node C, his Node would just have used that instead of creating a new one/
```

;;;    Carlos has a reference to Bob.  Reference possession is directional;
;;;    even though Alice has a reference to Bob, Bob does not have a
;;;    reference to Alice.
;;;
;;;  - One layer up is the "vat layer"... here we can see that Alice and
;;;    Alfred are both objects in Vat A, Bob is an object in Vat B, and
;;;    Carol and Carlos are objects in Vat C.
;;;
;;;  - Zooming out the farthest is the "node/network level".
;;;    There are two nodes (Node 1 and Node 2) connected over a
;;;    Goblins CapTP network.  The stubby shapes on the borders between the
;;;    nodes represent the directions of references Node 1 has to
;;;    objects in Node 2 (at the top) and references Node 2 has to
;;;    Node 1.  Both nodes in this diagram are cooperating to preserve
;;;    that Bob has access to Carol but that Carol does not have access to
;;;    Bob, and that Carlos has access to Bob but Bob does not have access
;;;    to Carlos.  (However there is no strict guarantee from either
;;;    node's perspective that this is the case... generally it's in
;;;    everyone's best interests to take a "principle of least authority"
;;;    approach though so usually it is.)
```

> The OCapN specifications leaves what happens after messages are delivered to the implementations and offers no opinions on what system might exist. OCapN has its roots in systems which implement an event loop called a *vat*. The vat has objects spawned in it and messages sent to these objects are queued up in a FIFO, the messages a delivered one at a time to the objects. Vats are transactional and if an error were to occur during message delivery, the transaction would be aborted leaving the vat in the prior state.

Reviewer: ```suggestion
The OCapN specifications leaves what happens after messages are delivered to the implementations and offers no opinions on what system might exist. OCapN has its roots in systems which implement an event loop plus object heap called a *vat*. The vat has objects spawned in it and messages sent to these objects are queued up in a FIFO, the messages a delivered one at a time to the objects. Vats are transactional and if an error were to occur during message delivery, the transaction would be aborted leaving the vat in the prior state.
```

E Vats were not transactional, I think only WaterKen Vats were.
